File: py/tests/_fixture.py
"""Shared LMDB fixture-writing helper for tests — writes raw entries via
plain lmdb.Environment, independent of the reader code under test."""

import lmdb


def write_fixture(objectbox_dir: str, entries: dict[bytes, bytes]) -> None:
    env = lmdb.Environment(
        objectbox_dir, map_size=10 * 1024 * 1024, max_dbs=4, readonly=False
    )
    with env.begin(write=True) as txn:
        for key, value in entries.items():
            txn.put(key, value)
    env.close()


def data_key(entity_id: int, object_id: int) -> bytes:
    return bytes([0x18, 0x00, 0x00, entity_id * 4]) + object_id.to_bytes(4, "big")


def relation_key(
    relation_id: int, direction: int, source_id: int, target_id: int
) -> bytes:
    return (
        bytes([0x08, 0x00, 0x00, (relation_id << 2) | direction])
        + source_id.to_bytes(4, "big")
        + target_id.to_bytes(4, "big")
    )


def schema_key(entity_id: int) -> bytes:
    return bytes([0x00, 0x00, 0x00, entity_id * 4]) + (0).to_bytes(4, "big")

File: js/test/fixture.ts
// Shared LMDB fixture-writing helper for tests — uses `lmdb` directly,
// independent of the reader code under test (mirrors py/tests/_fixture.py
// and dart/test/test_fixture.dart).
import { open } from "lmdb";

export function writeFixture(
  objectboxDir: string,
  entries: [Buffer, Buffer][],
): void {
  const db = open({
    path: objectboxDir,
    encoding: "binary",
    keyEncoding: "binary",
    mapSize: 10 * 1024 * 1024,
    // See src/index.ts's openStore for why this can't be left to lmdb-js's
    // own path-based guess.
    noSubdir: false,
  });
  for (const [key, value] of entries) {
    db.putSync(key, value);
  }
  db.close();
}

export function dataKey(entityId: number, objectId: number): Buffer {
  const key = Buffer.alloc(8);
  key[0] = 0x18; // type: object data
  key[3] = entityId * 4;
  key.writeUInt32BE(objectId, 4);
  return key;
}

export function relationKey(
  relationId: number,
  direction: number,
  sourceId: number,
  targetId: number,
): Buffer {
  const key = Buffer.alloc(12);
  key[0] = 0x08; // type: relation link
  key[3] = (relationId << 2) | direction;
  key.writeUInt32BE(sourceId, 4);
  key.writeUInt32BE(targetId, 8);
  return key;
}

export function schemaKey(entityId: number): Buffer {
  const key = Buffer.alloc(8);
  key[0] = 0x00; // type: schema entry
  key[7] = entityId;
  return key;
}
